src/formatting_stack/reporters/pretty_line_printer.clj

```diff
@@ -0,0 +1,79 @@
+(ns formatting-stack.reporters.pretty-line-printer
+  "Outputs report similar to github.com/xojs/xo"
+  (:require
+   [clojure.stacktrace :refer [print-stack-trace]]
+   [formatting-stack.protocols.reporter :as reporter]
+   [formatting-stack.protocols.spec :as protocols.spec]
+   [formatting-stack.reporters.impl :refer [truncate-line-wise]]
+   [formatting-stack.util :refer [colorize]]
+   [medley.core :refer [map-vals]]
+   [nedap.speced.def :as speced]
+   [nedap.utils.modular.api :refer [implement]]))
+
+(speced/defn print-summary [{:keys [^boolean? summary?
+                                    ^boolean? colorize?]} reports]
+  (when summary?
+    (->> reports
+         (group-by :level)
+         (map-vals count)
+         (into (sorted-map-by compare)) ;; print summary in order
+         (run! (fn [[report-type n]]
+                 (cond-> (str n (case report-type
+                                  :exception " exceptions occurred"
+                                  :error     " errors found"
+                                  :warning   " warnings found"))
+                   colorize? (colorize (case report-type
+                                         :exception :red
+                                         :error     :red
+                                         :warning   :yellow))
+                   true      println))))))
+
+(speced/defn print-exceptions [{:keys [^boolean? print-stacktraces?]} reports]
+  (->> reports
+       (filter (speced/fn [{:keys [^::protocols.spec/level level]}]
```

Comment:
Perhaps a bit late to start doing spec checks, would have personally defined a spec for `reports` I guess.

Also, lots of characters for `(comp #{:exception) :level)` ;).

Reply:
> Perhaps a bit late to start doing spec checks, would have personally defined a spec for reports I guess.


There's. the report-protocol actually checks this already.